Session2

Hyperbolic4183 opened the pull request:
## 行ったこと
reloadボタンが押された時に、fetchWeather()で返された値によって画像を変更されるようにした。

## 工夫したところ
- ビジネスロジックだけを取り出し、ビューを描画するファイルと切り分けた。


レビューよろしくお願いします！

Hyperbolic4183 committed:
画像を追加した。

Hyperbolic4183 committed:
fetchWeather()で取得した値を元に、天気の画像を表示させる処理を追加した。

daichikuwa0618 commented on Yumemi-ios-training/Assets.xcassets/sunny.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "sunny.png",
```
[課題](https://github.com/yumemi-inc/ios-training/blob/main/Documentation/API.md) を見ると、 pdf の画像が指定されていそうです。

ラスター画像とベクター画像ではちょっと扱いが違うので、できれば指定通り PDF で実装していただいていいですか？

Hyperbolic4183 replied:
@daichikuwa0618
.withRenderingModeで色を変えようとすると白の背景も変わってしまうので背景透過のものに変えてしまったんですよね。PDFのままの実装やってみます！

daichikuwa0618 commented on Yumemi-ios-training/WeatherDelegate.swift:
```diff
@@ -0,0 +1,12 @@
+//  Created by 大塚 周 on 2021/05/12.
+//
+
+import Foundation
```
[NITS]

`Foundation` は import 不要そうですね。

Hyperbolic4183 replied:
そうですね。消します。

daichikuwa0618 commented on Yumemi-ios-training/WeatherModel.swift:
```diff
@@ -0,0 +1,28 @@
+
+struct WeatherModel {
+    func reload(imageView: UIImageView)  {
+        print("test")
```
[NITS]

余計なデバッグ用出力は特に理由がない限り、極力削除して置きたいです。

Hyperbolic4183 replied:
わかりました。

daichikuwa0618 commented on Yumemi-ios-training/WeatherViewController.swift:
```diff
@@ -10,17 +10,25 @@ import UIKit
 
+extension WeatherViewController: WeatherDelegate {
+    func reload() {
+        weatherModel?.reload(imageView: weatherView.weatherImageView)
```
[IMO]

`weatherModel` という model 層のオブジェクトが `UIImageView` を操作している作りになっているので、修正が必要かもしれません。

(私なら) `weatherModel` からは天気だけを受け取って、 `WeatherViewController` で表示用のデータ (`UIImage` や `UIColor` を用意し、 `weatherView` に渡して View を更新させるって感じの処理にします。

他の方のアドバイスも聞きながら、考えてみてください。

一つの指標としては、何を import しているかは意識したほうが良いかも知れないです。
`WeatherModel` で `YumemiWeather` (仮想 API) と `UIKit` に依存していますが、これは外部 API と UI 層という離れた層に依存している形になっています。
UI 層オブジェクトであれば UI 関連だけ、 Model (Domain) 層であればロジック関連だけって感じの依存関係で完結するのが理想かと思います。

Hyperbolic4183 replied:
切り分けてみました。どうでしょうか？？

daichikuwa0618 commented on Yumemi-ios-training/WeatherDelegate.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol WeatherDelegate {
```
[IMO]
View の Interface を protocol で抽象化しているのはとても良いと思いました！！

せっかく抽象化を理解されているのであれば、 Model も積極的に抽象化する対象です。
理由としては UI よりもテストで出現する頻度が高く、stub/spy に置き換えられる場合が多いためです。

Session8 とかでテストコードは書くことになるので、ここではふーんと流しておいて、そのときに抽象化するでも大丈夫です 👌

daichikuwa0618 reviewed (changes requested):
いくつか diff をみて気になったことをレビューしました。

Hyperbolic4183 committed:
WeatherModel.swiftからUIKitに依存したロジックを削除した。
これまでweaterModelで行っていた画像の変更などのロジックをWeatherViewController.swiftとWeatherView.swiftで行うようにした。

Hyperbolic4183 committed:
画像を変更した。

daichikuwa0618 commented on Yumemi-ios-training/Assets.xcassets/cloudy.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "iconmonstr-weather-11.pdf",
+      "idiom" : "universal",
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
+      "scale" : "2x"
+    },
+    {
+      "idiom" : "universal",
+      "scale" : "3x"
```
[IMO]
pdf はベクター画像なので、特にラスター画像として扱いたい理由がなければそのままベクター画像として扱うのが一般的です。

- `Preserve Vector Data` を ON に
- Scales を `Single Scale` に

することでそのように扱えると思います。
(もし修正するのであれば 3 枚ともお願いします)

参考: https://qiita.com/okamu_/items/085cfc5999db801020ff

<img width="261" alt="スクリーンショット 2021-05-14 9 55 11" src="https://user-images.githubusercontent.com/31601805/118205061-88eb0900-b49a-11eb-9b0f-40a18d1b5e09.png">

Hyperbolic4183 replied:
なるほど、画質の悪い原因はこれだったんですね。

daichikuwa0618 commented on Yumemi-ios-training/WeatherDelegate.swift:
```diff
@@ -0,0 +1,10 @@
+//
+
+protocol WeatherDelegate {
+    func reload() -> Dictionary<String, Any>
```
[IMO]
ここは Dictionary で key が String だと型で縛れないので、表示する用のオブジェクトを作ってその型を返すようにするか、タプルで返すようにすることで型の制約をつけるのが良いと思います。

下記の例では `WeatherViewState` において `weather: String` と `color: UIColor` のプロパティを持っていますが、責務を意識できていれば画像をもたせたりしても良いかも知れないです。

<details><summary><b>型を作る時の例</b></summary><div>

```swift
struct WeatherViewState {
    let weather: String
    let color: UIColor
}

protocol WeatherDelegate {
    func reload() -> WeatherViewState
}
```

</div></details>

<details><summary><b>タプルで返す例</b></summary><div>

```swift
protocol WeatherDelegate {
    func reload() -> (weather: String, color: Color)
}
```

</div></details>

daichikuwa0618 reviewed (commented):
いくつか気になったのでレビューしました。

rizumita commented on Yumemi-ios-training/WeatherDelegate.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by 大塚 周 on 2021/05/12.
+//
+
+protocol WeatherDelegate {
```
## 命名
WeatherViewのDelegateなのでWeatherView…が良いように思います。データを取得する場合はUIKitでは一般的にDataSourceと命名されています。総合するとWeatherViewDataSourceとなります。

## Delegate内容
Delegateは処理の委譲ですがWeatherViewは「表示を行う」役割なのでreloadは役割として持っていないと考えられます。ですのでDelegateでreloadを委譲するということは無いように思います。

Hyperbolic4183 replied:
ありがとうございます！

Hyperbolic4183 committed:
Dictionary<String: Any>で天気の情報を返していたが、構造体でまとめた。

Hyperbolic4183 committed:
reloadをViewControllerのdelegateで行うものとした。

Hyperbolic4183 committed:
reloadをViewControllerで行うものとした。

Hyperbolic4183 commented:
@rizumita
いただいたレビューを反映させました。確認よろしくお願いします！

Hyperbolic4183 committed:
gitのキャッシュを消去

ruby109 commented on Yumemi-ios-training/WeatherViewController.swift:
```diff
@@ -9,6 +9,7 @@ import UIKit
 
 class WeatherViewController: UIViewController {
     
+    var delegate: WeatherViewDataSource?
```
reloadをViewControllerで行ってたなので、delegateのプロパティーを削除した方が良いと思います。

Hyperbolic4183 replied:
コメントありがとうございます。おっしゃる通り不必要なdelegateでした。

Hyperbolic4183 committed:
不必要なdelegateを消去した。
ファイルをフォルダに分けた。

Hyperbolic4183 committed:
WeatherModel -> WeatherDataSourceに変更した。

rizumita commented on Yumemi-ios-training/SceneDelegate.swift:
```diff
@@ -20,7 +20,9 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         if let windowScene = scene as? UIWindowScene {
             let window = UIWindow(windowScene: windowScene)
-            window.rootViewController = WeatherViewController()
+            let weatherViewController = WeatherViewController()
```
WeatherViewControllerのイニシャライザにweatherModelを渡す形にするとOptionalではなくする事が出来るように思います。

Hyperbolic4183 committed:
WeatherViewControllerのイニシャライザにモデルを加えた。

Hyperbolic4183 commented:
@Hyperbolic4183 あ

el-hoshino commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,46 @@
+
+class WeatherViewController: UIViewController {
+    
+    private(set) var weatherView = WeatherView()
```
[nits]
`weatherView` を別に後から変更するわけではないので、`let` でも一応問題ないです；
もしくは本来VCの生成コストを軽減するために生成時にビューの生成をなるべく控えることもありますが、そんな時に使うのは `private(set) lazy var` です

Hyperbolic4183 replied:
わかりました。

ykws commented on Yumemi-ios-training/Model/WeatherModel.swift:
```diff
@@ -0,0 +1,14 @@
+
+import YumemiWeather
+
+struct WeatherDataSource {
```
![badge](https://img.shields.io/badge/review-imo-orange.svg)

命名はファイル名の通り WeahterViewModel で良いと感じました。

xxxDataSource だと [UITableViewDataSource](https://developer.apple.com/documentation/uikit/uitableviewdatasource) protocol と同じような振る舞いを期待してしまうので。

CustomView に対して DataSource protocol を適合すると
次のようにやれなくもないですが、やり過ぎだと思うので参考程度に。

[CustomViewDataSourceExample](https://github.com/ykws/CustomDataSourceExample)

el-hoshino commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,46 @@
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    override func loadView() {
+        view = weatherView
+    }
```
👍 
VCのライフサイクル的に、`loadView` で `self.view` を設定することをどこかで勉強したんですか？

Hyperbolic4183 replied:
iOS設計パターン入門で勉強しました。ちょうど史さんのMVCのところですね。

ykws commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,46 @@
+            return WeatherViewState(weather: weatherString, color: .red)
+        case "cloudy":
+            return WeatherViewState(weather: weatherString, color: .gray)
+        default:
```
![badge](https://img.shields.io/badge/review-must-red.svg)

rainy と default を一緒に扱うのは避けたいです。
現状、 sunny, cloudy, rainy と提示されている以外を受け取っても表示できてしまうのも、サーバ側の不具合を握り潰しているのでプロダクト全体のバグの温床になりやすくなってしまっています。

Hyperbolic4183 replied:
わかりました。

el-hoshino commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,46 @@
+class WeatherViewController: UIViewController {
+    
+    private(set) var weatherView = WeatherView()
+    let weatherModel: WeatherDataSource
```
[nits]
まだSession2なので、責務の切り分けとかまだそこまで気にすることはないので、直接VC内で通信を組み込むのも現段階では特に問題ないですが、切り分けるなら、具象ではなく抽象に依存した方がテスト書きやすいのと、型名とプロパティ名を一致させた方が読みやすいです

ykws commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,46 @@
+    @objc func reloadButtonTapped() {
+        weatherView.changeDisplay(weatherViewState: reload())
+    }
+    func reload() -> WeatherViewState? {
```
![badge](https://img.shields.io/badge/review-ask-blue.svg)

ここは Optional が返る可能性ありますか？

Hyperbolic4183 replied:
ないですね。
これ以降のSessionでfetchWeathでエラーが返ってきた時にUIAlertVCとnilを返すのでつけていました。

el-hoshino commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,46 @@
+        super.viewDidLoad()
+        weatherView.reloadButton.addTarget(self, action: #selector(reloadButtonTapped), for: .touchUpInside)
+    }
+    @objc func reloadButtonTapped() {
```
実はこれは `tapped` とかの状況説明のメソッド名よりも、普通に `reload` とかの命令のメソッド名の方がいいです；
参考： https://qiita.com/bitpoetics/items/a63bd1d9afc0b3221890

ykws commented on Yumemi-ios-training/View/WeatherView.swift:
```diff
@@ -106,4 +106,10 @@ class WeatherView: UIView {
         reloadButton.setTitle("Reload", for: .normal)
     }
+    
+    func changeDisplay(weatherViewState: WeatherViewState?) {
```
![badge](https://img.shields.io/badge/review-imo-orange.svg)

changeDisplay なので、 Optional の受け取りを拒否して良いと思いました。

ykws commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,46 @@
+        weatherView.changeDisplay(weatherViewState: reload())
+    }
+    func reload() -> WeatherViewState? {
+        let weatherString = weatherModel.reload()
```
![badge](https://img.shields.io/badge/review-imo-orange.svg)

Model を経由して取得するなら WeatherViewState を返して欲しいところですね。
ViewController からロジックを取り除いて Model 側にあるのが理想です。

ykws commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,46 @@
+class WeatherViewController: UIViewController {
+    
+    private(set) var weatherView = WeatherView()
+    let weatherModel: WeatherDataSource
```
![badge](https://img.shields.io/badge/review-nits-green.svg)

別でコメントしていますが、こういうところで命名にブレが出てきたら再考の余地があるかもと思って見直してみてください。

Hyperbolic4183 replied:
わかりました。

el-hoshino commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,46 @@
+    @objc func reloadButtonTapped() {
+        weatherView.changeDisplay(weatherViewState: reload())
+    }
+    func reload() -> WeatherViewState? {
```
[must]
一般的には、副作用がなく、戻り値を返すメソッドは「コマンド」ではなく「クエリ」です；Swift API Design Guidelines的には、クエリメソッドは動詞の原形ではなく、過去形や現在進行形、もしくはそもそも動詞ではなく名詞や形容詞などを使った方がいいです（関数型パラダイムのメソッドには一部例外ありますが）
参考： https://swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage

ykws commented on Yumemi-ios-training/Model/WeatherViewState.swift:
```diff
@@ -0,0 +1,16 @@
+import UIKit
+
+struct WeatherViewState {
+    var weather: String
+    var color: UIColor
```
![badge](https://img.shields.io/badge/review-must-red.svg)

可変ではないと思うので let で良いかと

ykws commented on Yumemi-ios-training/Model/WeatherViewState.swift:
```diff
@@ -0,0 +1,16 @@
+struct WeatherViewState {
+    var weather: String
+    var color: UIColor
+    init(weather: String, color: UIColor) {
+        self.weather = weather
+        self.color = color
+    }
```
![badge](https://img.shields.io/badge/review-nits-green.svg)

property に対して set するだけなら initializer は省略可能です。

el-hoshino replied:
[nits]
structは自動的にmemberwise initializerが作られますので、自分で書かなくても大丈夫です

el-hoshino commented on Yumemi-ios-training/Model/WeatherModel.swift:
```diff
@@ -0,0 +1,14 @@
+import YumemiWeather
+
+struct WeatherDataSource {
+    func reload() -> String  {
```
[must]
これも同じく、コマンドではなくクエリです

ykws reviewed (changes requested):
ちゃんと View と Model を分ける意図があって良いと思います。
ただ、その責務の分離が明確になっていないように感じました。まだ小さいプログラムなので今の段階ではがんばって分けなくても大丈夫だったりはします。後半の Session では必須になりますが。
must の部分は見直し必須でお願いします！

Hyperbolic4183 committed:
breakの時にrainyとさせていたが、case: rainyを作成し、break: return nilとした。

Hyperbolic4183 commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -37,8 +37,10 @@ class WeatherViewController: UIViewController {
+        case "rainy":
             return WeatherViewState(weather: weatherString, color: .blue)
+        default:
+            return nil
```
@ykws 
これってdefaultでnil返す形が適切ですか？WeatherState型で返したいのですが、defaultを考慮するとオプショナルを加えざるをえなくなってしまうのですが、、、

ykws replied:
@Hyperbolic4183 良いつまずきです！
default を扱いたくない時は [Enumerations](https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html) の出番だと思うので、これを利用してデータを扱う方法にチャレンジしてみてください！

Hyperbolic4183 replied:
なるほど、enumはあるべくしてあるんですね。

Hyperbolic4183 committed:
名前を変更した。

Hyperbolic4183 committed:
イニシャライザを消去。

Hyperbolic4183 committed:
Weatherを列挙型としてを作成した。

Hyperbolic4183 committed:
クエリメソッドを動名詞に変更した。

Hyperbolic4183 committed:
ViewControllerからロジックを分離した。

Hyperbolic4183 committed:
Weather型の値からUIColorも定まるようなイニシャライザを作成した。

Hyperbolic4183 committed:
WeatherViewControllerのロジックを抽象に依存するようにした。

Hyperbolic4183 committed:
不必要なオプショナルを消去した。

Hyperbolic4183 commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -25,24 +25,25 @@ class WeatherViewController: UIViewController {
-        default:
-            return nil
         }
     }
```
@ykws 
列挙型Weatherを作成し、default でnilを返さなければならない問題を解決しました。
ただ、rawValueから返された値はオプショナル型なのでやむなく強制アンラップしていますが安全性などどうでしょうか？自分としてはここでのクラッシュの原因は一つに定まるので、良いと思っています。

ykws replied:
@Hyperbolic4183 `guard let weather` でエラー処理を追加するのが安全ですが、現段階では強制アンラップでクラッシュする可能性があることを認識していればこのままでも問題ないと思います。

Hyperbolic4183 replied:
@ykws
レビューありがとうございます！
`guard let `使う場合、 `else { return ○○}`で 何を返すことになりますか？
nilにした場合結局オプショナルをつけることになってしまい、WeatherViewState()にした場合はイニシャライズしないといけないと思いました。

ykws replied:
@Hyperbolic4183 すみません！　コメント見落としていました！
やるなら下記のコメントのような感じが私も良いと思います！

https://github.com/Hyperbolic4183/YumemiWeather/pull/4#discussion_r639348396

Hyperbolic4183 replied:
@ykws
いえいえ
クラッシュを回避したいことと何か返さなければいけないのでオプショナルを付与して長くなってしまうことのジレンマの解決方法がわかってよかったです。

Hyperbolic4183 commented on Yumemi-ios-training/View/WeatherView.swift:
```diff
@@ -16,6 +16,7 @@ class WeatherView: UIView {
     let highestTemperatureLabel = UILabel()
     let closeButton = UIButton(type: .system)
     let reloadButton = UIButton(type: .system)
+    var delegateAboutButton: WeatherDelegate?
```
@el-hoshino

- WeatherViewControllerからロジックをWeatherModelへと移した
~~- プロトコルを作成し、抽象に依存するよう工夫した~~
- 関数名を`reload`->`reloading`に変更した

先日いただいたレビューを元に変更を加えました。レビューよろしくお願いします！

ruby109 replied:
ロジックを抽象に依存することを意識するは良いです。
でも、`Delegate Pattern`を使わない場合、Delegateのプロパティは必要ではないと思います。

ruby109 commented on Yumemi-ios-training/Model/WeatherModel.swift:
```diff
@@ -13,11 +13,11 @@ struct WeatherModel {
         case .rainy:
-            return WeatherViewState(weather: .rainy, color: .blue)
+            return WeatherViewState(weather: .rainy)
         }
```
ここでは `return WeatherViewState(weather: weather)` で良いです

Hyperbolic4183 replied:
@ruby109
レビューありがとうございます！
いただいた２つのレビューを反映させましたがどうでしょうか？

Hyperbolic4183 committed:
プロトコルを消去した。

Hyperbolic4183 committed:
返り値を変数に変更した。

ruby109 commented on Yumemi-ios-training/Model/WeatherModel.swift:
```diff
@@ -13,11 +13,11 @@ struct WeatherModel {
+            return WeatherViewState(weather: weather)
         }
     }
 }
```
``` swift
let weather = Weather(rawValue: weatherString)!
return WeatherViewState(weather: weather)
```
Swiftなら、このような書き方で大丈夫です

Hyperbolic4183 replied:
@ruby109
知りませんでした！enumとswitchは相性いいですね。

Hyperbolic4183 committed:
switch文を簡潔にした。

ykws commented on Yumemi-ios-training/Model/Weather.swift:
```diff
@@ -0,0 +1,12 @@
+//
+
+enum Weather: String {
+    case sunny = "sunny"
+    case cloudy = "cloudy"
+    case rainy = "rainy"
```
![badge](https://img.shields.io/badge/review-nits-green.svg)

case と value が一致していたら省略可能です

```suggestion
    case sunny
    case cloudy
    case rainy
```

Hyperbolic4183 committed:
enumのvalueを省略した。

daichikuwa0618 commented on Yumemi-ios-training/View/WeatherView.swift:
```diff
@@ -106,4 +106,9 @@ class WeatherView: UIView {
     }
+    
+    func changeDisplay(weatherViewState: WeatherViewState) {
+        weatherImageView.image = UIImage(named: weatherViewState.weather.rawValue)?.withRenderingMode(.alwaysTemplate)
```
[FYI]
今回は大丈夫ですが、 `assets.xcassets` で `alwaysTemplate` の設定もできて、そこで実施した場合はコードで明示しなくても良くなります。

Hyperbolic4183 replied:
あ、そうなんですね。変更します。
preserve Vector Data ✅をAssets.xcassetsでやっているので、コードかAssets.xcassetsのどちらかに合わせた方が良さそうですね。

daichikuwa0618 commented on Yumemi-ios-training/Model/WeatherModel.swift:
```diff
@@ -0,0 +1,18 @@
+import YumemiWeather
+
+struct WeatherModel {
+    func reloading() -> WeatherViewState  {
```
[NITS]
`reload` でいいと思います。

daichikuwa0618 commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,34 @@
+//
+//  ViewController.swift
```
[NITS]
`WeatherViewController.swift` ですかね。

daichikuwa0618 reviewed (approved):
先を見越した責務の分離もできていて、 MVC パターンに則った設計にもできていると思います。
DI もできているのも素晴らしいです。

個人的にはマージして問題ないと思います。

Hyperbolic4183 committed:
xcAssetsでレンダリングモードを変更するようにした。

ykws reviewed (approved):
![LGTM](https://image.lgtmoon.dev/97014)

watanavex commented on Yumemi-ios-training/Controller/WeatherViewController.swift:
```diff
@@ -0,0 +1,34 @@
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
```
[FYI]
アノテーションを付与すると、このイニシャライザを呼び出している箇所をコンパイルエラーで検知できます。
こんな感じ↓
`@available(*, unavailable, message: "init(coder:) has not been implemented")`

watanavex commented on Yumemi-ios-training/Model/WeatherModel.swift:
```diff
@@ -0,0 +1,18 @@
+struct WeatherModel {
+    func reloading() -> WeatherViewState  {
+        let weatherString = YumemiWeather.fetchWeather()
+        let weather = Weather(rawValue: weatherString)!
```
[imo]
想定外の値でクラッシュさせる方針はアリだと思います！
その時、`fatalError` とかを使ってメッセージを出してあげると、
ログもわかりやすいし、コードの意図も伝わりやすいです。
(あえてやってるよ、という意図が伝わりますよね)
```Swift
guard let weather = Weather(rawValue: weatherString) else {
    fatalError("Unexpected xxx")
}

```

watanavex reviewed (approved):
些細なところコメントしましたが、このままマージでも良いと思います。
っていうか、完成度が高杉くん
![LGTM](https://i.lgtm.fun/1f9k.png)

watanavex commented on Yumemi-ios-training/View/WeatherView.swift:
```diff
@@ -106,4 +106,9 @@ class WeatherView: UIView {
     }
+    
+    func changeDisplay(weatherViewState: WeatherViewState) {
+        weatherImageView.image = UIImage(named: weatherViewState.weather.rawValue)
```
[imo]
rawValueとリソース名を一致させる　っという暗黙のルールは、ウッカリミスで壊れやすいです。
それならば、文字列リテラルの方が私は好みです。
R.swiftなどのライブラリを用いるのがベストですが。
まぁ、それもおいおいで良いかと思います。

Hyperbolic4183 replied:
レビューありがとうございます！リソース名の扱いについては色々方法がありますよね。
R.swiftはよく聞くので触ってみようと思います。
